torch/ao/quantization/pt2e/numeric_debugger.py

```diff
@@ -11,3 +20,188 @@ def generate_numeric_debug_handle(graph_module: GraphModule) -> None:
         if node.op != "placeholder" and NUMERIC_DEBUG_HANDLE_KEY not in node.meta:
             node.meta[NUMERIC_DEBUG_HANDLE_KEY] = unique_id
             unique_id += 1
+
+
+class OutputLogger(torch.nn.Module):
+    """
+    Base class for capturing output values.
+    """
+
+    # Mark as impure so that calls to it will not be removed during DCE.
+    _is_impure = True
+
+    def __init__(
+        self,
+        node_name: str,
+        nn_module_stack: object,
+        debug_handle: str,
+    ) -> None:
+        super().__init__()
+        self.node_name = node_name
+        self.nn_module_stack = nn_module_stack
+        self.debug_handle = debug_handle
+        self.stats: List[torch.Tensor] = []
+
+    def forward(self, x: object) -> object:
+        if isinstance(x, torch.Tensor):
+            self.stats.append(x.detach())
+        return x
+
+    def __repr__(self) -> str:
+        clean_dict = {
+            k: v
+            for k, v in self.__dict__.items()
+            # skip torch.nn.Module keys
+            if (k != "training") and not k.startswith("_")
+        }
+        return f"OutputLogger({clean_dict})"
+
+
+def _insert_logger(model: GraphModule, node: Node, debug_handle: str) -> None:
+    """For a given node, adds an OutputLogger that observes the output of that node,
+    and all its users use the OutputLogger output instead.
+    The OutputLogger will contain the debug_handle which can be used to compare
+    graphs after transforms"""
+
+    # to avoid circular dep
+    from torch.ao.quantization.fx.utils import get_new_attr_name_with_prefix
+
+    # add a logger after the node
+    with model.graph.inserting_after(node):
+        get_new_attr_name = get_new_attr_name_with_prefix(f"{node.name}_logger")
+        logger_name = get_new_attr_name(model)
+        setattr(
+            model,
+            logger_name,
+            OutputLogger(node.name, node.meta.get("nn_module_stack"), debug_handle),
+        )
+        logger_node = model.graph.call_module(logger_name, (node,), {})
+
+    orig_users = list(node.users.keys())
+    for user_node in orig_users:
+        if user_node is logger_node:
+            continue
+        user_node.replace_input_with(node, logger_node)
+
+
+def prepare_for_propagation_comparison(model: GraphModule) -> GraphModule:
+    # don't change the original model
+    model = copy.deepcopy(model)
+    for n in model.graph.nodes:
+        if NUMERIC_DEBUG_HANDLE_KEY not in n.meta:
+            continue
+        numeric_debug_handle = n.meta[NUMERIC_DEBUG_HANDLE_KEY]
+        _insert_logger(model, n, numeric_debug_handle)
+
+    model.recompile()
+    return model
+
+@dataclass(frozen=True)
+class QuantizationComparisonResult:
+    actual: torch.Tensor
+    ref: torch.Tensor
+
```

Comment:
I recommend adding `__post_init__` and assert that the actual/ref are really tensors